src/System.Private.CoreLib/src/System/String.cs

```diff
@@ -775,95 +767,120 @@ private unsafe static bool StartsWithOrdinalHelper(String str, String startsWith
 
         private unsafe static int CompareOrdinalHelper(String strA, String strB)
         {
+            Contract.Requires(strA != null);
+            Contract.Requires(strB != null);
+
+            // NOTE: This may be subject to change if eliminating the check
+            // in the callers makes them small enough to be inlined
+            Contract.Assert(strA._firstChar == strB._firstChar,
+                "For performance reasons, callers of this method should " +
+                "check/short-circuit beforehand if the first char is the same.");
+
             int length = Math.Min(strA.Length, strB.Length);
-            int diffOffset = -1;
 
             fixed (char* ap = &strA._firstChar) fixed (char* bp = &strB._firstChar)
             {
                 char* a = ap;
                 char* b = bp;
 
-                // unroll the loop
-                while (length >= 10)
-                {
-                    if (*(int*)a != *(int*)b)
-                    {
-                        diffOffset = 0;
-                        break;
-                    }
-
-                    if (*(int*)(a + 2) != *(int*)(b + 2))
-                    {
-                        diffOffset = 2;
-                        break;
-                    }
+                // Check if the second chars are different here
+                // The reason we check if _firstChar is different is because
+                // it's the most common case and allows us to avoid a method call
+                // to here.
+                // The reason we check if the second char is different is because
+                // if the first two chars the same we can increment by 4 bytes,
+                // leaving us word-aligned on both 32-bit (12 bytes into the string)
+                // and 64-bit (16 bytes) platforms.
+        
+                // For empty strings, the second char will be null due to padding.
+                // The start of the string is the EE type pointer + string length,
+                // which takes up 8 bytes on 32-bit, 12 on x64. For empty strings,
+                // the null terminator immediately follows, leaving us with an object
+                // 10/14 bytes in size. Since everything needs to be a multiple
+                // of 4/8, this will get padded and zeroed out.
+                
+                // For one-char strings the second char will be the null terminator.
```

Comment:
Please let me know if there are any inconsistencies in this comment block; I'm assuming the CoreRT string layout is mostly the same as the one in CoreCLR.